### PR TITLE
Add better error messages for bad colormaps

### LINKED ISSUE #2400: Pad colors when not all control points specified
## To Reproduce

```python
import numpy as np

import napari

colormap = napari.utils.Colormap(colors=[[0,0,0,1], [1, 1, 1, 1]], name="test", controls=[0.1, 0.9] )

print(colormap.colorbar)

with napari.gui_qt():
    viewer = napari.Viewer()
    viewer.add_image(np.random.random(size=(100, 100)), colormap=colormap)
```

end of stacktrace
```
~/.pyenv/versions/3.8.3/envs/napari/lib/python3.8/site-packages/vispy/color/colormap.py in _glsl_mix(controls=array([0.1, 0.9], dtype=float32), colors=array([[0., 0., 0., 1.],
    143 
    144     """
--> 145     assert (controls[0], controls[-1]) == (0., 1.)
```

## Expected behavior

If Colormap is properly constructed (and `colorbar` works) then it should not raise an exception when applying to layer. 

## Possible solutions

1. Fix controls and colors before passing to vispy (napari/napari/_vispy/vispy_image_layer.py line 128) by adding a copy of first and last color with respectively coordinate 0 and 1 if absent.
2. Raise an exception if trying to create a colormap with first and last coordinate different to (0, 1)

I prefer the first solution.

## Environment
last master (42d48e3d63aa850ec18a3a6167e7441275c94303)

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR closes #2400 and supersedes and closes #2422 by providing better error messages for "bad" colormaps. It adds more comprehensive tests to make sure the errors are raised in the right way.

Those issues were discussed at a recent team meeting and folks generally thought that raising more clear errors was a better approach than guessing at the correct behavior. Let us know @Czaki if this PR is ok with you. Thanks for your work on the other ones, I think it helped us come to this decision more clearly.

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

# References

Closes #2400
Closes #2422

# How has this been tested?

- [x] example: Tests added

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/utils/colormaps/_tests/test_colormap.py [hunk 1]
```diff
@@ -27,6 +28,38 @@ def test_linear_colormap_with_control_points():
...
+
+def test_wrong_start_control_point():
+    """Test wrong start of control points raises an error."""
+    colors = np.array([[0, 0, 0, 1], [0, 1, 0, 1], [0, 0, 1, 1]])
+    with pytest.raises(ValueError):
+        Colormap(colors, name='testing', controls=[0.1, 0.75, 1])
+
+
+def test_wrong_end_control_point():
+    """Test wrong start of control points raises an error."""
```
Comment (kevinyamauchi) [inline napari/utils/colormaps/_tests/test_colormap.py]:
```suggestion
    """Test wrong end of control points raises an error."""
```

**Review by kevinyamauchi — APPROVED** [review]
Thanks, looks good to me, @sofroniewn ! I agree that it is better to give a clear error message than try to automagically fix it. I made one minor suggestion below, but otherwise I think this is good to go.

**Comment by Czaki** [comment]
If you think that translation of exceptions is a good idea then provide more verbose information in exception (list of problematic points for example).

File: napari/utils/colormaps/_tests/test_colormap.py [hunk 2]
```diff
@@ -27,6 +28,38 @@ def test_linear_colormap_with_control_points():
     np.testing.assert_almost_equal(cmap.map([0.75]), [[0, 1, 0, 1]])
 
 
+def test_non_ascending_control_points():
+    """Test non ascending control points raises an error."""
```
Comment (andy-sweet) [inline napari/utils/colormaps/_tests/test_colormap.py:32]: Is there a benefit to adding the docstring here and in other tests? To me the name of the test and its body are self-descriptive and read well, so the docstring doesn't seem to add much and I'd remove it.

I see docstrings for tests being more useful when the test is complex or when it refers to an issue (e.g. on GitHub), but I might be missing something.

File: napari/utils/colormaps/colormap.py [hunk 3]
```diff
@@ -67,11 +67,43 @@ def _ensure_color_array(cls, v):
...
+            )
+
+        # Check number of control points is correct
+        n_controls_target = len(values['colors']) + int(
+            values['interpolation'] == ColormapInterpolationMode.ZERO
+        )
+        if not len(v) == n_controls_target:
+            raise ValueError(
+                trans._(
+                    'Wrong number of control points provided',
```
Comment (andy-sweet) [inline napari/utils/colormaps/colormap.py]: Optional: might be useful to add something like

`f'Expected {n_controls_target}, got {n_controls}'`

**Review by andy-sweet — APPROVED** [review]

File: napari/utils/colormaps/_tests/test_colormap.py (hunk 2 above)
Comment (sofroniewn) [reply napari/utils/colormaps/_tests/test_colormap.py:32]: it's a very valid question - i think we have a non-uniform adoption of doc strings in our tests. some are longer, some are shorter. i agree this one doesn't help much, but i might still be inclined to leave for now as i don't see it hurting.

if someone else also weighed in to say we should drop them then i'd happily do so. i don't have strong feelings here at all though. we could do an audit of our test docs strings at some point too and standardize things/ clean them up

File: napari/utils/colormaps/colormap.py (hunk 3 above)
Comment (sofroniewn) [reply napari/utils/colormaps/colormap.py]: will do

**Comment by sofroniewn** [comment]
> If you think that translation of exceptions is a good idea then provide more verbose information in exception

Done.

Thanks for reviews here, will merge after tests pass!

**Pull request merged**